foreign_cc/nmake.bzl

```diff
@@ -0,0 +1,57 @@
+"""Helper rules and macros for building projects using the MSVC nmake build tool"""
+
+load("//foreign_cc:providers.bzl", "ForeignCcDepsInfo")
+load(":configure.bzl", "configure_make")
+
+def _extra_toolchains_transition_impl(settings, attrs):
+    return {"//command_line_option:extra_toolchains": attrs.extra_toolchains + settings["//command_line_option:extra_toolchains"]}
+
+_extra_toolchains_transition = transition(
+    implementation = _extra_toolchains_transition_impl,
+    inputs = ["//command_line_option:extra_toolchains"],
+    outputs = ["//command_line_option:extra_toolchains"],
+)
+
+def _extra_toolchains_transitioned_foreign_cc_target_impl(ctx):
+    # Return the providers from the transitioned foreign_cc target
+    return [
+        ctx.attr.target[DefaultInfo],
+        ctx.attr.target[CcInfo],
+        ctx.attr.target[ForeignCcDepsInfo],
+        ctx.attr.target[OutputGroupInfo],
+    ]
+
+extra_toolchains_transitioned_foreign_cc_target = rule(
+    doc = "A rule for adding extra toolchains to consider when building the given target",
+    implementation = _extra_toolchains_transitioned_foreign_cc_target_impl,
+    cfg = _extra_toolchains_transition,
+    attrs = {
+        "extra_toolchains": attr.string_list(
+            doc = "Additional toolchains to consider",
+            mandatory = True,
+        ),
+        "target": attr.label(
+            doc = "The target to build after considering the extra toolchains",
+            providers = [ForeignCcDepsInfo],
+            mandatory = True,
+        ),
+        "_allowlist_function_transition": attr.label(
+            default = "@bazel_tools//tools/allowlists/function_transition_allowlist",
+        ),
+    },
+    incompatible_use_toolchain_transition = True,
+)
+
+def configure_nmake(name, **kwargs):
+    """ Wrapper macro around configure_make to force usage of the preinstalled nmake toolchain """
+    configure_make_target_name = name + "_nmake"
+
+    configure_make(
+        name = configure_make_target_name,
+        **kwargs
+    )
+    extra_toolchains_transitioned_foreign_cc_target(
+        name = name,
+        extra_toolchains = ["@rules_foreign_cc//toolchains:preinstalled_nmake_toolchain"],
```

Comment:
I feel like this should be configurable. Maybe instead of having a dedicated `configure_nmake` target, there's a `configure_make_variant` rule? And you can pass in a toolchain you want the build to use?

The functionality for this should then be moved to `/foreign_cc/private` and setup for `make` and `configure_make`.... what do you think?

Reply:
Hmm I could make it more generic, although I'm not aware of any `make` variants other than GNU Make and NMake. If that is unlikely to ever change then perhaps having `configure_make` and `configure_nmake` is clearer and easy to use.

Plus `rules_foreign_cc` allow the registering of custom native toolchains via the [`rules_foreign_cc_dependencies`](https://github.com/bazelbuild/rules_foreign_cc/blob/main/foreign_cc/repositories.bzl#L9) macro. 

If a user must use a `configure_make_variant` and provide a toolchain to use, it will be confusing which toolchain will be used, the one passed in to `configure_make_variant` or the one passed into `rules_foreign_cc_dependencies`?

With `configure_make` and `configure_nmake` rules, this confusion is abstracted away from the user.

Hope that made sense :)

Reply:
Looking at [wikipedia](https://en.wikipedia.org/wiki/Make_(software)#Derivatives) there seem to be quite a few. I don't think I see how the explicit `configure_nmake` vs `configure_make_variant` makes a difference in how clear it is what toolchain is being used. The core issue I'm worried about is someone wanting to define a custom `nmake` toolchain but is not able to get the rule to use it because we only transition to add the one built into `rules_foreign_cc`. It could be the case that someone uses [jom](https://wiki.qt.io/Jom) in their project (a derivative of `nmake`) but has no way to get the rules to use it.

Reply:
Ah ok that all makes sense, I didn't realise how many variants there are.

To clarify, `configure_make` should be replaced with `configure_make_variant`, which includes an attribute that allows  the user to specify the toolchain. Something like:

```
configure_make_variant(
 name = "openssl_nmake"
 variant_toolchain = "@rules_foreign_cc//toolchains:preinstalled_nmake_toolchain"
)

configure_make_variant(
 name = "openssl_gnumake"
 variant_toolchain = "@rules_foreign_cc//toolchains:prebuilt_make_toolchain"
)

or should `configure_make` continue to exist?
```

Reply:
Oh, I somehow lost my response....

I think `configure_make` should continue to exists. The new `*_variant` macros should only be considered special case helpers for people who have cross platform dependencies that use weird toolchains. I would say that the way the toolchains are currently setup are incorrect. If Bazel is using a `msvc` compiler, it should probably be using `nmake`. But I think this would be a larger breaking change so to transition things over, the `*_variant` rules help the user explicitly define the toolchain they want.

But just to be clear, I think the goal is to have users define the toolchains the traditional way and the rules should continue to try to support that. Replacing rules which do traditional toolchain management vs this custom solution should be avoided if possible. But I think this is a justifiable case until we can revisit how to manage these toolchains.

What do you think?

Reply:
Sure, sounds good to me! I’ll try to find some time to implement what you suggest 😁👍🏻

Reply:
You're a legend!! 🙌